### DISCUSSION TITLE
Support native enums newtypes

### DISCUSSION THREAD

**Discussion by SamuelMarks** [diesel-rs/diesel#4368] [Ideas]
Here is my `up.sql` (simplified without duplicate_object exception handling):
```sql
-- https://www.healthline.com/health/beauty-skin-care/skin-undertones#choosing-your-color-palette
CREATE TYPE skin_undertone AS ENUM
    ('cool', 'warm', 'neutral', 'olive');

-- https://www.schemecolor.com/real-skin-tones-color-palette.php
CREATE TYPE skin_colour AS ENUM
    ('russet', 'peru', 'fawn', 'mellow_apricot', 'navajo_white');

-- https://tsemrtd.js.org/enums/consts_enums.HairColor.html
CREATE TYPE hair_colour AS ENUM (
    'bald', 'black', 'blond', 'brown',
    'gray', 'white', 'auburn', 'other');
```

Diesel generates this:
```rs
pub mod sql_types {
    #[derive(diesel::query_builder::QueryId, Clone, diesel::sql_types::SqlType)]
    #[diesel(postgres_type(name = "hair_colour"))]
    pub struct HairColour;

    #[derive(diesel::query_builder::QueryId, Clone, diesel::sql_types::SqlType)]
    #[diesel(postgres_type(name = "skin_colour"))]
    pub struct SkinColour;

    #[derive(diesel::query_builder::QueryId, Clone, diesel::sql_types::SqlType)]
    #[diesel(postgres_type(name = "skin_undertone"))]
    pub struct SkinUndertone;
}
```

Trying to get it to work, added `Debug, PartialEq`, got most errors to disappear, but then getting errors like:
```
the trait `ToSql<SkinColour, Pg>` is not implemented for `SkinColour`
```

Found this example from 2018 https://github.com/diesel-rs/diesel/issues/1783#issuecomment-412470766 - so can probably hack around there until it works (mapping to a real newtype rather than `smallint`). Something like:
```rs
use strum_macros;

#[derive(diesel::AsExpression, diesel::FromSqlRow, strum::EnumString, PartialEq, Debug, Clone)]
#[strum(serialize_all = "snake_case")]
#[sql_type = "crate::schema::models::skin_undertone"]
pub enum SkinUndertone {
    Cool = 0,
    Warm = 1,
    Neutral,
    Olive,
}

#[derive(diesel::AsExpression, diesel::FromSqlRow, strum::EnumString, PartialEq, Debug, Clone)]
#[strum(serialize_all = "snake_case")]
#[sql_type = "crate::schema::models::skin_colour"]
pub enum SkinColour {
    Russet = 0,
    Peru = 1,
    Fawn,
    MellowApricot,
    NavajoWhite,
}

#[derive(diesel::AsExpression, diesel::FromSqlRow, PartialEq, Debug, Clone)]
#[strum(serialize_all = "snake_case")]
#[sql_type = "crate::schema::models::hair_colour"]
pub enum HairColour {
    Bald = 0,
    Black = 1,
    Blond,
    Brown,
    Gray,
    White,
    Auburn,
    Other, // = -1,
}
```

Any chance we can have native enum support in diesel?

FWIW: SQLalchemy in Python has gotten this to work - https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.Enum

Thanks for your consideration

**Comment by YievCkim** [comment]
Hi,

I am trying to work with native PG enum too. And I don't get the examples. it seems these constructs try to translate Rust enum type to PG integer.

What is strum_macros in your example ?

**Reply by SamuelMarks** [reply]
https://crates.io/crates/strum_macros

Basically a nice way of going from string to enum and from enum to string [+ related things].

**Reply by YievCkim** [reply]
Cool I am going to take a look. Thanks !

**Reply by YievCkim** [reply]
Nice crate. it's something I often need. Thank you for that !

**Comment by weiznich** [comment]
There is native support for enum types in diesel. We just expect you to provide your own rust enums (like we also do for other types). You obviously need to implement the corresponding type mappings, so in this case `FromSql` and `ToSql`. I won't provide anymore details, due to the ping on discord, but the documentation on this is quite complete.

**Reply by SamuelMarks** [reply]
Apologies for the ping.
